packages/wallet/src/rollup-state-validator.ts

```diff
@@ -0,0 +1,285 @@
+      const swapper: Address = snapshots[0].state.pubKey
+      const convertedTx: Swap = {
+        sender: swapper,
+        tokenType: transition.tokenType as UniTokenType | PigiTokenType,
+        inputAmount: transition.inputAmount,
+        minOutputAmount: transition.minOutputAmount,
+        timeout: transition.timeout,
+      }
+      return {
+        signature: transition.signature,
+        transaction: convertedTx,
+      }
+    }
+
+    return undefined
+  }
+
+  public async checkNextTransition(
+    nextTransition: RollupTransition
+  ): Promise<FraudCheckResult> {
+    let preppedFraudInputs: StateSnapshot[]
+    let generatedPostRoot: Buffer
+
+    const transitionPostRoot: Buffer = hexStrToBuf(nextTransition.stateRoot)
+
+    if (isCreateAndTransferTransition(nextTransition)) {
+      const slotIfSequential: number = await this.rollupMachine.getNextNewAccountSlot()
+      // if the created slot is not sequential, for now it will break
+      if (slotIfSequential < nextTransition.recipientSlotIndex) {
+        throw new AggregatorUnsupportedError()
+      }
+    }
+
+    // In case there was fraud in this transaction, get state snapshots for each input so we can prove the fraud later.
+    preppedFraudInputs = await this.getInputStateSnapshots(nextTransition)
+    // let inputAsTransaction: SignedTransaction = await this.getTransactionFromTransition(nextTransition)
+    const inputAsTransaction: SignedTransaction = await this.getTransactionFromTransitionAndSnapshots(
+      nextTransition,
+      preppedFraudInputs
+    )
+    try {
+      await this.rollupMachine.applyTransaction(inputAsTransaction)
+      generatedPostRoot = await this.rollupMachine.getStateRoot()
+    } catch (error) {
+      if (isStateTransitionError(error)) {
+        // return the fraud proof, invalid transaction
+        return {
+          fraudPosition: this.currentPosition,
+          fraudInputs: preppedFraudInputs,
+          fraudTransition: nextTransition,
+        }
+      } else {
+        throw new LocalMachineError()
+      }
+    }
+
+    if (generatedPostRoot.equals(transitionPostRoot)) {
+      this.currentPosition.blockNumber++
+      this.currentPosition.transitionIndex++
+      return 'NO_FRAUD'
+    } else {
+      // return the fraud proof, invalid root
+      return {
+        fraudPosition: this.currentPosition,
+        fraudInputs: preppedFraudInputs,
+        fraudTransition: nextTransition,
+      }
+    }
+  }
+
+  public async checkNextBlock(
+    nextBlock: RollupBlock
+  ): Promise<any> {
+    const currentPosition: RollupTransitionPosition = await this.getCurrentVerifiedPosition()
+
+    if (nextBlock.blockNumber !== currentPosition.blockNumber + 1) {
+      throw new ValidationOutOfOrderError()
+    }
+
+    for (const transition of nextBlock.transitions) {
+      const fraudCheck: FraudCheckResult = await this.checkNextTransition(
+        transition
+      )
+      if (fraudCheck !== 'NO_FRAUD') {
```

Comment:
Can change `checkNextTransition(...)` to `getTransitionFraudProof(...)`, and then `if (!!fraudProof)` means there was fraud and you won't have to cast it to `LocalFraudProof` below.